enhancements/ingress/optional-ingress.md

```diff
@@ -0,0 +1,249 @@
+
+#### Authentication operator
+
+No change to the authentication operator is needed until the implementation of the ingress capability for the standalone OpenShift.
+HyperShift doesn't use the authentication operator.
+
+#### Monitoring operator
+
+No change to the monitoring operator is needed until the implementation of the ingress capability for the standalone OpenShift.
+The monitoring operator skips the alert manager's route check if `ingress` cluster operator is not there.
+
+### Risks and Mitigations
+
+#### Partially implemented capability
+
+The ingress capability will become a part of the capability API. Meanwhile its implementation will be incomplete due to the missing support of the standalone OpenShift. However the usage of the capability is an important stepping stone, it unifies the implementation of the optional ingress for all the OpenShift installation types.
+
+#### Component routes
+
+In the absence of the ingress operator, the component routes cannot be customized by the owner components due to the lack of the RBAC.
+The cluster administrator can customize a component route with a custom TLS certificate secret but the component owner won't be able to access the secret. This may be misleading for the cluster administrators. However the component routes were never fully implemented on HyperShift (authentication operator is not used).
+
+### Drawbacks
+
+In the absence of the default ingress controller, the Route API is left unserved. The OpenShift components or workloads can create routes but no controller is available to back them up. This may be misleading for the users.
+
+## Design Details
+
+### Open Questions [optional]
+
+N/A
+
+### Test Plan
+TBD
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+N/A
+
+#### Tech Preview -> GA
+N/A
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+
+Same as [the component selection strategy](/enhancements/installer/component-selection.md#upgrade--downgrade-strategy).     
+Note that:
+- the capabilities [cannot disabled post install](/enhancements/installer/component-selection.md#capabilities-cannot-be-uninstalled)
+- the capabilities can be [implicitly enabled](/enhancements/installer/component-selection.md#updates) (e.g. cluster version with `None` baseline upgrades to the version which moves some of the core payloads to a capability)
+
+### Version Skew Strategy
+N/A
+
+### Operational Aspects of API Extensions
+N/A - no api extensions are being introduced in this proposal
+
+#### Failure Modes
+N/A - no api extensions are being introduced in this proposal
+
+#### Support Procedures
+N/A - no api extensions are being introduced in this proposal
+
+## Implementation History
+N/A
+
+## Alternatives
```

Comment:
two more alternatives/modifications:

1) just allow standalone clusters to be broken if Ingress is not enabled.  This would mean anyone using the `None` capset would need to also explicitly add the Ingress as an additionalEnabledCapability.  This is somewhat like the current proposal where the installer would validate/require/auto-inject the additional cap.  The downside is a questionable UX(users find out when their cluster doesn't come up that their install config wasn't valid) and user confusion about why the cap is defined if it can't be disabled.  The upside is there isn't as much hidden magic going on under the covers enabling caps, magic which a user might accidentally or deliberately circumvent by, for example, rendering the cluster resources and then modifying the ClusterVersion object prior to creating the cluster.

2) courtesy of @dhellmann - can we make use of cluster profiles for now instead?  Hypershift already has a defined clusterprofile (hosted-control-plane or some such name) as does standalone OCP, so can we just label the resources into the right profiles, rather than using the capability feature for this, for now?  Or is there an expectation that some hypershift clusters will need/want to turn the ingress-operator on?

Reply:
> Hypershift already has a defined clusterprofile (hosted-control-plane or some such name) as does standalone OCP, so can we just label the resources into the right profiles, rather than using the capability feature for this, for now?

I'm still catching up on recent changes here, but we don't actually have a HyperShift profile at the moment:

```console
$ oc adm release extract --to manifests quay.io/openshift-release-dev/ocp-release:4.14.0-ec.1-x86_64
$ grep -ohr 'include.release.openshift.io/.*' manifests | sort | uniq -c
    620 include.release.openshift.io/ibm-cloud-managed: "true"
      7 include.release.openshift.io/ibm-cloud-managed: 'true'
    980 include.release.openshift.io/self-managed-high-availability: "true"
      8 include.release.openshift.io/self-managed-high-availability: 'true'
    832 include.release.openshift.io/single-node-developer: "true"
      7 include.release.openshift.io/single-node-developer: 'true'
```

The `ibm-cloud-managed` profile is ROKS, and it's HyperShift-like, but  not HyperShift.  `single-node-developer` is legacy cruft, as described in openshift/cluster-version-operator#685, and `self-managed-high-availability` is standalone OpenShift.  HyperShift currently has [this][1], which they hope to replace with a profile eventually, but haven't yet.  And there isn't currently an in-cluster API for "what profile am I?", outside of the cluster-version operator's Deployment `spec`, which isn't visible in the hosted Kubernetes API.

> Or is there an expectation that some hypershift clusters will need/want to turn the ingress-operator on?

Seems like "allow some HyperShift clusters to still deploy the ingress component" should be either a goal or a non-goal? 

[1]: https://github.com/openshift/hypershift/blob/203ffe95ed50591ccd059e1a6bbd638bd2014ab7/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L47-L50

Reply:
last i knew "ibm-cloud-managed" represented the resource that hypershift deploys.  so while it's not called the hypershift profile, effectively it is.  Perhaps things have evolved and hypershift uses a different mechanism to filter resources now, but that's how it used to work.

> Seems like "allow some HyperShift clusters to still deploy the ingress component" should be either a goal or a non-goal?


yeah, needs clarity.

Reply:
> just allow standalone clusters to be broken if Ingress is not enabled. 

Yes, this can be an alternative indeed. I like the fact that there will be no need to tweak the installer's code. This deserves to be written down as an alternative but I'm wondering whether this can be the main approach as it seems to be even more questionable UX than the "implicit adding/validation error".

Reply:
> courtesy of @dhellmann - can we make use of cluster profiles for now instead? Hypershift already has a defined clusterprofile (hosted-control-plane or some such name) as does standalone OCP, so can we just label the resources into the right profiles, rather than using the capability feature for this, for now? Or is there an expectation that some hypershift clusters will need/want to turn the ingress-operator on?

As far as I understand:
1. The capability is the expected way to compose an OpenShift instance on HyperShift. For instance, the console operator is supposed to be enabled/disabled using the already implemented capability.
2. HyperShift is supposed to be enhanced with the capability API which, I suppose again, would look similar to the CVO/install-config one. Meaning that it would allow the users to turn the ingress on (not only disable).

@csrwng , @enxebre : can you help checking the points above?

Reply:
@alebedev87 Yes, agree on both points. The idea is to not do away with ingress completely, but make it optional when creating your cluster.